Enable fp32/bf16 PRelu forward and backward in MkldnnCPU path

yanbing-j opened the pull request:
Enable fp32/bf16 PRelu forward and backward in MkldnnCPU path.

Fixes https://github.com/pytorch/pytorch/issues/58896

yanbing-j committed:
Enable fp32/bf16 PRelu forward and backward in MkldnnCPU path

yanbing-j commented:
@XiaobingSuper @jingxu10

yanbing-j commented:
@mingfeima

xsacha commented:
Thanks. This helps a lot!

jingxu10 commented:
@xsacha There are some minor changes compared to the one that I shared to you before, like the following 2.
1. added weight shape check in aten/src/ATen/native/mkldnn/PRelu.cpp
2. some edits in tools/build_variables.bzl and torch/_C/_nn.pyi.in

xsacha commented:
I've been using this for a while now (in libtorch) and it's working well. Thankyou.
Avoids repeated conversion to and from mkldnn layout within my models.

VitalyFedyunin commented:
Im about half-through review this PR, and wonder if you really need to update ideep here

VitalyFedyunin commented on test/test_mkldnn.py:
```diff
@@ -354,6 +354,70 @@ def test_relu_bf16(self):
+        x_bf16 = x.clone().to_mkldnn(torch.bfloat16).requires_grad_()
+        m = mkldnn_utils.to_mkldnn(torch.nn.PReLU())
+        m_bf16 = mkldnn_utils.to_mkldnn(torch.nn.PReLU(), torch.bfloat16)
+        if has_bf16_support():
```
nit: this check can be at the begining of the test

yanbing-j replied:
Done.

jingxu10 commented:
PReLU is a primitive in oneDNN, we need to enable it in ideep for the invocation from PyTorch. Thus, we need to update ideep as well.
@yanbing-j @XiaobingSuper

VitalyFedyunin commented:
Please update ideep as separate PR, so we can match it with internal update.

yanbing-j commented:
> Please update ideep as separate PR, so we can match it with internal update.

@VitalyFedyunin The PR of updating ideep is https://github.com/pytorch/pytorch/pull/61696. Please review it. Thanks.

mingfeima commented:
@yanbing-j , sorry i was on leave last week.

I don't think we need MKLDNN for `PReLU` here. Basically we should limit mkldnn to the operator that are difficult or impossible to be efficiently enabled by PyTorch native manner, aka. "ATen". `PReLU` is a fairly simple and we can achieve best perf with ATen for sure. Just do manual vectorization with vec256 will do the job.

Benefits:
* will favor other platforms as well not just only Intel CPU;
* more generic, e.g. "non-contiguous" cases.

This is different from gelu which involves Gaussian distribution which is the very difficult to implement it in ATen manner, but PReLU is not the case. So insert a kernel from `cpu/Activation.cpp` with `TensorIterator` is a more proper way of doing it.

jingxu10 commented:
@mingfeima any drawbacks if both of oneDNN PReLU and vec256 are enabled? It won't be a bad idea to keep this PR as what it is to leave an interface to oneDNN PReLU if it is needed in the future.

yanbing-j commented:
> @mingfeima any drawbacks if both of oneDNN PReLU and vec256 are enabled? It won't be a bad idea to keep this PR as what it is to leave an interface to oneDNN PReLU if it is needed in the future.

@jingxu10

yanbing-j commented:
The content of changing third_party/ideep has been merged by #63748. This PR has been rebased on the latest master.

yanbing-j committed:
Fix Lint-flake8 error

yanbing-j commented:
Hi @VitalyFedyunin , could you please help review this PR? Customer cares about this usage.

ngimel commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -3691,15 +3691,26 @@
     CPU: prelu_backward_cpu
     CUDA: prelu_backward_cuda
 
+- func: mkldnn_prelu(Tensor self, Tensor weight) -> Tensor
```
why do we need these functions? MkldnnPrelu can dispatch regular `prelu` to mkldnn

yanbing-j replied:
Hi @ngimel , these functions are for module `MkldnnPrelu` to find the correct functions when doing `to_mkldnn()`. https://github.com/pytorch/pytorch/blob/e3fc20f676d97348c02647295b2dbf3aca878a2f/torch/utils/mkldnn.py#L200
https://github.com/pytorch/pytorch/blob/e3fc20f676d97348c02647295b2dbf3aca878a2f/test/test_mkldnn.py#L409

ngimel replied:
Those functions should be correctly selected during dispatch on mkldnn tensors, they shouldn't be separate functions in native_functions.yaml

yanbing-j replied:
Got it. I have removed `mkldnn_prelu` in native_functions.yaml. And use `torch.prelu` to call the function, and it can be correctly selected during dispatch.

yanbing-j committed:
Remove unnecessary mkldnn_prelu dispatch

yanbing-j commented:
@ngimel Could you please help review this PR? Thanks.

ngimel commented:
I'm getting a lot of internal errors like
```
stderr: caffe2/aten/src/ATen/native/mkldnn/Prelu.cpp:44:3: error: no member named 'prelu_forward' in namespace 'ideep'; did you mean 'dnnl::prelu_forward'?
  ideep::prelu_forward::compute(
```
that I don't know how to fix. cc @VitalyFedyunin. 
Why do we need this PR and shouldn't previous improvement for prelu to use vectorized path be enough?

xsacha commented:
> Why do we need this PR and shouldn't previous improvement for prelu to use vectorized path be enough?

Would the vectorised CPU path work on a tensor that is of type Mkldnn? Or is there a conversion cost in both directions, considering the rest of the model is expected to be converted to Mkldnn weights.

ngimel commented:
Prelu is a pointwise operation, so one alternative would be to enable dispatch to vectorized implementation for mkldnn tensors, it should produce correct results regardless of the data layout.

xsacha commented:
Thanks. That's an acceptable alternative if it's too much to merge this in.

yanbing-j commented:
Hi @ngimel, Could you please provide your ideep commit id? We will check ideep version. And I also rebase this branch again. Thanks!

ngimel commented:
Sorry, I don't know ideep commit id, and I wouldn't know how to update it internally. cc @VitalyFedyunin

VitalyFedyunin commented:
Ideep should be up-to-date right now, what you need to do is remove ideep update from this PR and rebase to master.

yanbing-j commented:
> Ideep should be up-to-date right now, what you need to do is remove ideep update from this PR and rebase to master.

Hi @VitalyFedyunin , I have already removed ideep update from this PR and rebased it to master.

mingfeima commented:
@yanbing-j please rebase this one.

yanbing-j committed:
Merge branch 'master' into prelu_mkldnn

yanbing-j commented:
> @yanbing-j please rebase this one.

Done.

xsacha commented:
Is this ready for merge? I saw the CPU version of this PR is already merged.

yanbing-j commented:
> Is this ready for merge? I saw the CPU version of this PR is already merged.

It is waiting for the reviewer to land this PR.

yanbing-j committed:
Merge branch 'master' into prelu_mkldnn

yanbing-j commented:
Hi @VitalyFedyunin @malfet @ngimel ,could you please help land this PR? Thanks!
